docs/adr/0007-preventing-version-mismatches.md

```diff
@@ -0,0 +1,233 @@
+    change `uniffi` or `uniffi-bindgen`.
+
+This means that our current method of using exact string matching for versions is no longer going to work.
+
+Finally, there can only be one `uniffi-bindgen` and `uniffi-bindgen-[language]`, installed system wide.  This makes it
+difficult for a developer to work on crates that depend on different UniFFI versions.
+
+We've built a decent solution for the scaffolding code: [the builtin-bingen
+feature](https://mozilla.github.io/uniffi-rs/tutorial/Rust_scaffolding.html#avoiding-version-mismatches-between-uniffi-core-and-uniffi-bindgen).
+However, there's no similar solution for the bindings code.
+
+application-services has a custom solution for the bindings generation.  It contains the
+[embedded-uniffi-bindgen](https://github.com/mozilla/application-services/tree/main/tools/embedded-uniffi-bindgen) crate
+which contains very
+[simple binary](https://github.com/mozilla/application-services/blob/main/tools/embedded-uniffi-bindgen/src/main.rs)
+that just runs `uniffi-bindgen`. This allows us to control the `uniffi-bindgen` version using the `Cargo.toml` for this
+crate, rather than relying on the system installed version.
+
+### Breaking changes to the UDL syntax
+
+A related issue is breaking changes to the UDL syntax, for example the change the renamed `[Wrapped]` to `[Custom]`.
+These kind of changes can lead to challenges with scaffolding and bindings generation.  For example, users could get
+into a situtation where `uniffi-bindgen` can parse a UDL file, but `uniffi-bindgen-kotlin` can't.
+
+However, this issue seems less important since a) the errors are easier to diagnose and fix and b) we have historically
+only introduced breaking changes like this for "experimental" features like custom/wrapped types.
+
+This ADR will ignore this issue and only focus on compatibility between the generated scaffolding and bindings code.
+
+## Decision Drivers
+
+- UniFFI should have a friendly developer experience.
+  - If possible, developers shouldn't have to worry or know about this issue.
+  - If that's not possible, then it should be easy for them to identify and fix the issue.
+- We need to support external bindings generators and allow those projects to
+  release new versions without coordinating with the UniFFI team.
+
+## Considered Options
+
+### **[Requirement] More flexible versioning**
+
+As mentioned in the context, once we have external bindings generators we can't rely on a simply compare it's version
+string to `uniffi`.  A more flexible versioning policy is a requirement for all other options discussed here.
+
+I think we can use semver for this:
+  - The major version is incremented whenever there's a change that would breaks compatibility between the bindings and
+    scaffolding generation code. Changing the layout of `RustBuffer` or `RustCallStatus` would be a breaking change. Our
+    custom/external types work would not.
```

Comment:
To make this workable, I think we'd need to generate an explicit list of the things that can cause a breaking change (Probably we'd fail to enumerate all of them on a first attempt, but it's worth attempting anyway).

Reply:
That's a very good point, and I wonder if identifying this might help to get to the crux of what we are trying to solve here.